### PR TITLE
Support options such as --help and --version without a config file

### PR DESCRIPTION
**Pull request by yallop**
Fixes mirage/mirage#472.

### Before

```
$ mirage help
[ERROR] No configuration file config.ml found.
Please specify the configuration file using -f.
```
```
$ mirage --version
[ERROR] No configuration file config.ml found.
Please specify the configuration file using -f.
```
```
$ mirage help config
[ERROR] No configuration file config.ml found.
Please specify the configuration file using -f.
```

### After

```
$ mirage help
MIRAGE(1)                        Mirage Manual                       MIRAGE(1)



NAME
       mirage - The mirage application builder
[...]
```
```
$ mirage --version
2.6.1
```
```
$ mirage help config
MIRAGE-CONFIGURE(1)              Mirage Manual             MIRAGE-CONFIGURE(1)



NAME
       mirage-configure - Configure a mirage application.
[...]
```

### CODE REVIEW

**Comment by Drup** [comment]
In the current kv_ro example (not the functoria version).

```
% mirage --help
Fatal error: exception Invalid_argument("Base key map is not available at this point. Please stop messing with functoria's invariants.")
```

**Comment by yallop** [comment]
@Drup: I think the `Base key map` problem is fixed now.

**Comment by Drup** [comment]
Well, it's fixed, but the documentation doesn't work at all anymore for `mirage configure --help`.

**Comment by Drup** [comment]
I think the base key issue comes from one of your earlier refactoring.

**Comment by yallop** [comment]
I've pushed a revised version (c94c90b) in which all of `mirage --help`, `mirage --version`, `mirage configure --help` and `mirage help` do something reasonable regardless of whether there's a config file available, and which doesn't fall over with pre-functoria config files such as `kv_ro`.

File: app/functoria_app.ml [hunk 1]
```diff
@@ -612,8 +614,28 @@ module Make (P: S) = struct
...
+        argv
+    in
+    match result with
+    | `Error _ -> exit 1
+    | `Ok Cmd.Help -> ()
+    | `Ok (Cmd.Configure _ | Cmd.Describe _ | Cmd.Build _ | Cmd.Clean _) ->
+      Functoria_misc.Log.fatal "%s" error
+    | `Version
+    | `Help -> ()
+    with Invalid_argument _ -> assert false
```
Comment (Drup) [inline app/functoria_app.ml]: Is that a leftover before you changed it into a proper exception ?

File: app/functoria_app.ml [hunk 2]
```diff
@@ -573,10 +572,13 @@ module Make (P: S) = struct
...
-    registered () >>= fun t ->
-    Log.set_section (Config.name t);
-    Ok t
+    try
+      compile_and_dynlink file >>= fun () ->
+      registered () >>= fun t ->
+      Log.set_section (Config.name t);
+      Ok t
+    with Config'.No_base_key_map ->
+      Error "Access to base key map is unavailable at this point"
```
Comment (Drup) [inline app/functoria_app.ml]: Maybe we should specify which "point" we are talking about ? Something like "Access to base key map is not available before configuration. For dynamically configured devices, please use the configurable class." ?

File: app/functoria_app.ml (hunk 1 above)
Comment (yallop) [inline app/functoria_app.ml]: Indeed.  I've removed it ( 5ec12cb).

File: app/functoria_app.ml (hunk 2 above)
Comment (yallop) [inline app/functoria_app.ml]: Good idea: fixed (c0ab272d).

**Comment by Drup** [comment]
Nice! I tested, it works well. Thanks!

**Pull request merged**